File: modules/image-palette/android/src/main/java/expo/modules/imagepalette/ImagePaletteModule.kt
package expo.modules.imagepalette

import android.util.Log
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ImagePaletteModule : Module() {

  private var config = Config()
  private var firstReadOccurred = false
  private val paletteCache = PaletteCache()

  private fun log(msg: String) {
    Log.d(TAG, "[ImagePalette] $msg")
  }

  override fun definition() = ModuleDefinition {
    Name("ImagePalette")

    OnDestroy {
      paletteCache.clear()
      config = Config()
      firstReadOccurred = false
    }

    Function("configure") { options: PaletteConfigOptions ->
      if (firstReadOccurred) {
        log("configure() ignored: must be called before the first getDominantColors() call")
        return@Function
      }
      val warnings = config.apply(options)
      warnings.forEach { log("configure: $it — ignored") }
    }

    AsyncFunction("getDominantColors") { uri: String ->
      firstReadOccurred = true
      val cfg = config.copy()
      log("start")

      val key = PaletteKey.fromConfig(uri, cfg)
      if (cfg.cache) {
        paletteCache.get(key)?.let {
          log("cache HIT")
          return@AsyncFunction respond(it, cfg)
        }
        log("cache MISS")
      }

      val path = if (uri.startsWith("file://")) uri.removePrefix("file://") else uri

      val bitmap = ImageDecoder.decode(path, cfg.downsample, cfg.downsampleTargetSize)
        ?: return@AsyncFunction null

      val decoded = ImageDecoder.extractARGB(bitmap)

      val swatches = HistogramQuantizer.quantize(
        decoded,
        cfg.gridWidth,
        cfg.gridHeight,
        cfg.edgesOnly,
        cfg.bitsPerChannel
      )

      if (cfg.cache) {
        paletteCache.set(key, swatches)
      }

      log("resolved")
      respond(swatches, cfg)
    }
  }

  private fun respond(swatches: List<InternalSwatch>, config: Config): Map<String, Any> {
    return mapOf(
      "gridWidth" to config.gridWidth,
      "gridHeight" to config.gridHeight,
      "swatches" to swatches.map {
        mapOf(
          "row" to it.row,
          "col" to it.col,
          "r" to it.r,
          "g" to it.g,
          "b" to it.b,
          "population" to it.population
        )
      }
    )
  }

  companion object {
    private const val TAG = "ImagePalette"
  }
}

File: modules/image-palette/ios/ImagePaletteModule.swift
import ExpoModulesCore
import UIKit
import os

public class ImagePaletteModule: Module {

  private var config = Config()
  private var firstReadOccurred = false
  private let paletteCache = PaletteCache()

  private static let logger = Logger(subsystem: "com.imagepalette", category: "main")

  private func log(_ msg: String) {
    let line = "[ImagePalette] \(msg)"
    Self.logger.info("\(line, privacy: .public)")
  }

  public func definition() -> ModuleDefinition {
    Name("ImagePalette")

    OnDestroy {
      self.paletteCache.clear()
      self.config = Config()
      self.firstReadOccurred = false
    }

    Function("configure") { (options: PaletteConfigOptions) -> Void in
      if self.firstReadOccurred {
        self.log("configure() ignored: must be called before the first getDominantColors() call")
        return
      }
      let warnings = self.config.apply(options)
      for warning in warnings {
        self.log("configure: \(warning) — ignored")
      }
    }

    AsyncFunction("getDominantColors") { (uri: String) -> [String: Any]? in
      self.firstReadOccurred = true
      let cfg = self.config
      self.log("start")

      let key = PaletteKey(uri: uri, config: cfg)
      if cfg.cache, let cached = self.paletteCache.get(key) {
        self.log("cache HIT")
        return self.respond(swatches: cached, config: cfg)
      }
      if cfg.cache {
        self.log("cache MISS")
      }

      let path = uri.hasPrefix("file://") ? String(uri.dropFirst(7)) : uri

      guard let image = ImageDecoder.decode(
              path: path,
              downsample: cfg.downsample,
              targetSize: cfg.downsampleTargetSize
            ) else {
        return nil
      }

      guard let decoded = ImageDecoder.extractRGBA(image: image) else {
        return nil
      }

      let swatches = HistogramQuantizer.quantize(
        decoded: decoded,
        gridWidth: cfg.gridWidth,
        gridHeight: cfg.gridHeight,
        edgesOnly: cfg.edgesOnly,
        bitsPerChannel: cfg.bitsPerChannel
      )

      if cfg.cache {
        self.paletteCache.set(key, swatches)
      }

      self.log("resolved")
      return self.respond(swatches: swatches, config: cfg)
    }
  }

  private func respond(swatches: [InternalSwatch], config: Config) -> [String: Any] {
    return [
      "gridWidth": config.gridWidth,
      "gridHeight": config.gridHeight,
      "swatches": swatches.map { swatch in
        [
          "row": swatch.row,
          "col": swatch.col,
          "r": swatch.r,
          "g": swatch.g,
          "b": swatch.b,
          "population": swatch.population
        ] as [String: Any]
      }
    ]
  }
}
